### PR TITLE
Add must-gather-output-dir option

### PR DESCRIPTION
**Pull request by chetna14manku**
Add must-gather incase addon/operator install fails

### CODE REVIEW

File: ocp_addons_operators_cli/utils/addons_utils.py [hunk 1]
```diff
@@ -120,6 +124,15 @@ def assert_addons_user_input(addons, brew_token):
         assert_missing_managed_odh_brew_token(addons=addons, brew_token=brew_token)
 
 
+def store_kubeconfig_from_cluster_name(cluster_name, kubeconfig_dict):
```
Comment (rnetser) [inline ocp_addons_operators_cli/utils/addons_utils.py:127]: cluster-object has `kubeconfig`
pass the object and use `cluster.name` and `cluster.kubeconfig`

File: ocp_addons_operators_cli/utils/addons_utils.py [hunk 2]
```diff
@@ -143,6 +156,9 @@ def prepare_addons(addons, ocm_token, endpoint, brew_token, install):
             client=ocm_client,
             name=cluster_name,
         )
+        addon["kubeconfig"] = store_kubeconfig_from_cluster_name(
```
Comment (rnetser) [inline ocp_addons_operators_cli/utils/addons_utils.py:159]: move under `if cluster.exists:`

File: ocp_addons_operators_cli/utils/cli_utils.py [hunk 3]
```diff
@@ -46,6 +49,7 @@ def verify_user_input(**kwargs):
     addons = kwargs.get("addons")
     ocm_token = kwargs.get("ocm_token")
     brew_token = kwargs.get("brew_token")
+    kwargs.get("must_gather_output_dir")
```
Comment (rnetser) [inline ocp_addons_operators_cli/utils/cli_utils.py:52]: why is this needed here?

File: ocp_addons_operators_cli/utils/cli_utils.py [hunk 4]
```diff
@@ -82,10 +88,24 @@ def run_install_or_uninstall_products(operators, addons, parallel, debug, instal
...
+                    LOGGER.error(
+                        f"{product_dict['name']} Install Failed. "
+                        f"Running must-gather for {product_dict['cluster_name']}"
+                    )
+                    collect_must_gather(
+                        must_gather_output_dir,
+                        product_dict["kubeconfig"],
+                        product_dict["cluster_name"],
+                        product_dict["name"],
+                    )
```
Comment (rnetser) [inline ocp_addons_operators_cli/utils/cli_utils.py:108]: use arg=val

File: ocp_addons_operators_cli/utils/cli_utils.py [hunk 5]
```diff
@@ -82,10 +88,24 @@ def run_install_or_uninstall_products(operators, addons, parallel, debug, instal
...
-                futures.append(executor.submit(action_func, **action_kwargs))
-            else:
-                processed_results.append(action_func(**action_kwargs))
+            product_dict = product_action_tuple[2]
+            try:
+                if parallel:
+                    futures.append(executor.submit(action_func, **action_kwargs))
+                else:
+                    processed_results.append(action_func(**action_kwargs))
+            except Exception:
```
Comment (rnetser) [inline ocp_addons_operators_cli/utils/cli_utils.py:97]: how does it work in parallel?

File: ocp_addons_operators_cli/utils/general.py [hunk 6]
```diff
@@ -89,3 +96,40 @@ def get_iib_dict():
...
+
+
+def collect_must_gather(
+    must_gather_output_dir, kubeconfig_path, cluster_name, product_name
+):
+    try:
+        target_dir = os.path.join(
+            must_gather_output_dir, "must-gather", cluster_name, product_name
+        )
+    except Exception as ex:
```
Comment (myakove) [inline ocp_addons_operators_cli/utils/general.py:108]: The error seems unrelated (you try to get target dir, not exec must gather here)
also this will never fail, it is just join()

File: ocp_addons_operators_cli/utils/cli_utils.py [hunk 7]
```diff
@@ -82,10 +88,24 @@ def run_install_or_uninstall_products(operators, addons, parallel, debug, instal
         for product_action_tuple in addons_action_list + operators_action_list:
             action_func = product_action_tuple[0]
             action_kwargs = product_action_tuple[1]
-            if parallel:
-                futures.append(executor.submit(action_func, **action_kwargs))
-            else:
-                processed_results.append(action_func(**action_kwargs))
+            product_dict = product_action_tuple[2]
+            try:
+                if parallel:
```
Comment (myakove) [inline ocp_addons_operators_cli/utils/cli_utils.py:93]: This need to be converted to be like we have in `openshift-cli-installer`
https://github.com/RedHatQE/openshift-cli-installer/blob/main/openshift_cli_installer/libs/clusters/ocp_clusters.py#L172

**Review by myakove — CHANGES_REQUESTED** [review]

File: ocp_addons_operators_cli/utils/general.py (hunk 6 above)
Comment (chetna14manku) [reply ocp_addons_operators_cli/utils/general.py:108]: I took help from the `openshift-cli-installer` [code](https://github.com/RedHatQE/openshift-cli-installer/blob/main/openshift_cli_installer/libs/clusters/ocp_cluster.py#L245), now I understand it doesnot check the path, instead function will create this path itself.

File: ocp_addons_operators_cli/utils/addons_utils.py (hunk 1 above)
Comment (rnetser) [reply ocp_addons_operators_cli/utils/addons_utils.py:110]: this is not resolved

File: ocp_addons_operators_cli/utils/addons_utils.py [hunk 8]
```diff
@@ -104,7 +107,15 @@ def assert_addons_user_input(addons, brew_token):
         assert_missing_managed_odh_brew_token(addons=addons, brew_token=brew_token)
 
 
-def prepare_addons(addons, ocm_token, endpoint, brew_token, install):
+def store_kubeconfig_from_cluster_name(cluster_name, kubeconfig_dict):
```
Comment (rnetser) [inline ocp_addons_operators_cli/utils/addons_utils.py:110]:
```suggestion
def write_kubeconfig_file(cluster_name, kubeconfig_dict):
```

File: ocp_addons_operators_cli/utils/addons_utils.py [hunk 9]
```diff
@@ -104,7 +107,15 @@ def assert_addons_user_input(addons, brew_token):
...
+    kubeconfig_path = os.path.join("/tmp", f"kubeconfig-{cluster_name}")
+    with open(kubeconfig_path, "w") as fd:
+        fd.write(yaml.dump(kubeconfig_dict))  # noqa: FCN001
```
Comment (rnetser) [inline ocp_addons_operators_cli/utils/addons_utils.py:113]: add 'yaml' to fcn_exclude_functions

File: ocp_addons_operators_cli/utils/addons_utils.py [hunk 10]
```diff
@@ -168,6 +183,10 @@ def prepare_addons_action(addons, install):
             brew_token = addon.get("brew-token")
             if brew_token:
                 action_kwargs["brew_token"] = brew_token
+            must_gather_output_dir = addon.get("must_gather_output_dir")
+            if must_gather_output_dir:
```
Comment (rnetser) [inline ocp_addons_operators_cli/utils/addons_utils.py:187]:
```suggestion
            if must_gather_output_dir := addon.get("must_gather_output_dir"):
```

File: ocp_addons_operators_cli/utils/operators_utils.py [hunk 11]
```diff
@@ -127,6 +128,11 @@ def prepare_operators_action(operators, install):
             action_kwargs["iib_index_image"] = operator.get("iib")
             action_kwargs["source_image"] = operator.get("source-image")
             action_kwargs["target_namespaces"] = operator.get("target-namespaces")
+            must_gather_output_dir = operator.get("must_gather_output_dir")
+            if must_gather_output_dir:
```
Comment (rnetser) [inline ocp_addons_operators_cli/utils/operators_utils.py:132]: ditto

**Review by rnetser — CHANGES_REQUESTED** [review]

File: ocp_addons_operators_cli/utils/addons_utils.py (hunk 9 above)
Comment (chetna14manku) [reply ocp_addons_operators_cli/utils/addons_utils.py:113]: @rnetser it is already added in .flake8 file https://github.com/RedHatQE/ocp-addons-operators-cli/blob/main/.flake8#L15

File: ocp_addons_operators_cli/utils/addons_utils.py (hunk 1 above)
Comment (chetna14manku) [reply ocp_addons_operators_cli/utils/addons_utils.py]: updated the function

File: ocp_addons_operators_cli/utils/addons_utils.py (hunk 9 above)
Comment (chetna14manku) [reply ocp_addons_operators_cli/utils/addons_utils.py]: added `write` to the file and it resolved the pre-commit failure.

**Review by rnetser — APPROVED** [review]

**Review by rnetser — APPROVED** [review]

**Review by rnetser — DISMISSED** [review]

**Review by rnetser — APPROVED** [review]

**Review by myakove — APPROVED** [review]
